File: agent/dotnet/src/Perper.Protocol/Cache/Instance/StreamData.cs
using System.Collections;

using Apache.Ignite.Core.Binary;

namespace Perper.Protocol.Cache.Instance
{
    public static class StreamData
    {

        public static IBinaryObjectBuilder Create(
            IBinary binary,
            string agent,
            string instance,
            string @delegate,
            StreamDelegateType delegateType,
            bool ephemeral,
            object[] parameters,
            string? indexType = null,
            Hashtable? indexFields = null)
        {
            var streamData = binary.GetBuilder($"StreamData_{agent}_{@delegate}");

            streamData.SetField("agent", agent);
            streamData.SetField("instance", instance);
            streamData.SetField("delegate", @delegate);
            streamData.SetField("delegateType", delegateType);
            streamData.SetField("ephemeral", ephemeral);
            streamData.SetField("listeners", new ArrayList());
            streamData.SetField("indexFields", indexFields);
            streamData.SetField("indexType", indexType);
            streamData.SetField("parameters", parameters);

            return streamData;
        }

        public static IBinaryObjectBuilder AddListener(
            IBinaryObjectBuilder streamData,
            IBinaryObject listener)
        {
            var listeners = streamData.GetField<ArrayList>("listeners");
            listeners.Add(listener.ToBuilder());
            streamData.SetField("listeners", listeners);

            return streamData;
        }

        public static IBinaryObjectBuilder RemoveListener(
            IBinaryObjectBuilder streamData,
            IBinaryObject listener)
        {
            var listeners = streamData.GetField<ArrayList>("listeners");
            listeners.Remove(listener);
            streamData.SetField("listeners", listeners);

            return streamData;
        }

        public static IBinaryObjectBuilder RemoveListener(
            IBinaryObjectBuilder streamData,
            string caller,
            int parameter)
        {
            var listeners = streamData.GetField<ArrayList>("listeners");
            for (var i = 0 ; i < listeners.Count ; i++)
            {
                var listener = (IBinaryObject)listeners[i]!;
                if (listener.GetField<string>("caller") == caller && listener.GetField<int>("parameter") == parameter)
                {
                    listeners.RemoveAt(i);
                    break;
                }
            }
            streamData.SetField("listeners", listeners);

            return streamData;
        }
    }
}

File: agent/python/perper/protocol/stream_data.py
from collections import OrderedDict
from pyignite import GenericObjectMeta
from pyignite.datatypes.complex import ObjectArrayObject
from pyignite.datatypes import String, IntObject, BoolObject, MapObject, EnumObject, CollectionObject

class StreamData(metaclass=GenericObjectMeta, schema=OrderedDict([
    ('instance', String),
    ('agent', String),
    ('delegate', String),
    ('delegateType', EnumObject),
    ('parameters', ObjectArrayObject),

    ('ephemeral', BoolObject),
    ('indexType', String),
    ('indexFields', MapObject),

    ('listeners', CollectionObject),
])):
    pass

class StreamListener(metaclass=GenericObjectMeta, schema=OrderedDict([
    ('callerAgent', String),
    ('callerInstance', String),
    ('caller', String),
    ('parameter', IntObject),
    ('filter', MapObject),
    ('replay', BoolObject),
    ('localToData', BoolObject),
])):
    pass
